packages/modules/psa/api.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+
+import logging
+from typing import Optional
+
+from modules.common import req
+from typing import NamedTuple, Tuple
+
+from modules.psa.config import PSAConfiguration
+
+log = logging.getLogger("soc."+__name__)
+
+ManufacturerConfiguration = NamedTuple(
+    "ManufacturerConfiguration",
+    [("brand", str), ("realm", str), ("client_id", Optional[str]), ("client_secret", Optional[str])]
+)
+
+manufacturer_configurations = {
+    "Peugeot": ManufacturerConfiguration(
+        brand="peugeot.com",
+        realm="clientsB2CPeugeot",
+        client_id="1eebc2d5-5df3-459b-a624-20abfcf82530",
+        client_secret="T5tP7iS0cO8sC0lA2iE2aR7gK6uE5rF3lJ8pC3nO1pR7tL8vU1",
+    ),
+    "Citroen": ManufacturerConfiguration(
+        brand="citroen.com",
+        realm="clientsB2CCitroen",
+        client_id="5364defc-80e6-447b-bec6-4af8d1542cae",
+        client_secret="iE0cD8bB0yJ0dS6rO3nN1hI2wU7uA5xR4gP7lD6vM0oH0nS8dN",
+    ),
+    "DS": ManufacturerConfiguration(
+        brand="driveds.com",
+        realm="clientsB2CDS",
+        client_id="cbf74ee7-a303-4c3d-aba3-29f5994e2dfa",
+        client_secret="X6bE6yQ3tH1cG5oA6aW4fS6hK0cR0aK5yN2wE4hP8vL8oW5gU3",
+    ),
+    "Opel": ManufacturerConfiguration(
+        brand="opel.com",
+        realm="clientsB2COpel",
+        client_id="07364655-93cb-4194-8158-6b035ac2c24c",
+        client_secret="F2kK7lC5kF5qN7tM0wT8kE3cW1dP0wC5pI6vC0sQ5iP5cN8cJ8",
+    ),
+    "Vauxhall": ManufacturerConfiguration(
+        brand="vauxhall.co.uk",
+        realm="clientsB2CVauxhall",
+        client_id=None,  # don't have valid client_id and secret for Vauxhall
+        client_secret=None,  # don't have valid client_id and secret for Vauxhall
+    )
+}
+
+
+def fetch_access_token(user_id: str, password: str, client_id: Optional[str], client_secret: Optional[str],
+                       manufacturer: str, session: req.Session
+                       ) -> Tuple[str, str, str]:
```

Comment:
Die Aufteilung ist so viel besser, aber die schiere Anzahl an Parametern und Rückgabewerten finde ich unschön.

Hier ein Gegenvorschlag:
```python
def create_session(user_id: str, password: str, client_id: Optional[str], client_secret: Optional[str],
                   manufacturer: str) -> Session:
    manufacturer_config = manufacturer_configurations[manufacturer]
    brand = manufacturer_config.brand
    realm = manufacturer_config.realm
    if not client_id or not client_secret:  # use defaults if no client_id and client_secret is specified
        client_id = manufacturer_config.client_id
        client_secret = manufacturer_config.client_secret
        if not client_id or not client_secret:
            raise Exception(
                "No OAuth credentials configured and no default available for manufacturer %s" % manufacturer)

    session = req.get_http_session()
    data = {
        'realm': realm, 'grant_type': 'password', 'password': password, 'username': user_id, 'scope': 'openid profile'
    }

    access_token = session.post(
        "https://idpcvs." + str(brand) + "/am/oauth2/access_token", data=data, auth=(client_id, client_secret)
    ).json()['access_token']

    session.params = {"client_id": client_id}
    session.headers = {
        'Accept': 'application/hal+json', 'Authorization': 'Bearer %s' % access_token, 'x-introspect-realm': realm
    }
    return session
```
Ungetestet, müsste aber meine ich funktionieren und du hast damit alle Parameter für die ganze Session gesetzt und kannst diese auch in den anderen Funktionen einsparen.

Reply:
ja, das ist in der Tat schöner, danke! Ist gestestet und funktioniert. PR ist aktualisiert.